rfc-143-split-database-instances.md

```diff
@@ -0,0 +1,66 @@
+
+
+## Naming
+
+The word "database" is often interchangeably used to describe a collection of tables, and an instance of (e.g.) Postgres server running somewhere. For the purposes of this RFC, "database" refers to the collection of tables, schema, etc.. We'll use "instance" to refer to the runtime that hosts the database.
+
+## Problem
+
+Having a single instance hosting multiple databases has a number of drawbacks:
+
+1. Major upgrades must be applied to all databases in one go, as it's the instance we're upgrading, not the database. Major upgrades are considered risky because they contain potentially breaking, and backward-incompatible changes. This means major upgrades are scarier, so we're less likely to do them until we're forced. They also require coordination across disparate teams so everyone's confident their apps will survive the upgrade.
+
+2. Problems in one database are likely to impact other unrelated services. For example an app with unexpectedly high load on its database could max-out the instance, causing performance degradation in other databases and their apps.
+
+3. Related to [2], it's significantly harder to size instance resources. Disk is fairly straightforward, but it's harder to understand CPU/RAM/IOPS patterns. We're also likely to over-size resources to reduce the likelihood large-scale problems in [2].
+
+4. There is no clear owner of the instances. Each app is owned by a single team, and each database should be accessed by exactly one app. However, many apps with different owners reside on the single instance. Sizing and managing databases is an app-level responsibility, so it doesn't fit with the remit of Platform Reliability/Replatforming/Platform Health (i.e. for a single, central resource).
+
+5. We still host some database features ourselves (such as Mongo 2.6) which has a maintenance burden and allows us to run very out of date versions of these services. 
+
+This is coming up now because our versions of MySQL and Postgres are approaching End of Life (notice for [MySQL](https://forums.aws.amazon.com/ann.jspa?annID=8790) and [Postgres](https://forums.aws.amazon.com/ann.jspa?annID=8499)), forcing us to plan major upgrades for both over the next few months.
+
+## Proposal
+
+We propose a new principle to be applied across GOV.UK:
+
+1. Supporting services for an application should not be shared with any other applications at the infrastructure/provisioning level.
+
+> Note: this principle doesn't cover services used for application communication, e.g. where SNS/SQS/RabbitMQ broadcasts events for other applications. The use of the service carries most weight here – e.g. a fictional RabbitMQ queue used solely to manage Sidekiq jobs would be considered a "supporting service", whereas a RabbitMQ queue to broadcast publishing domain events would be accessed by many apps. Both these queues would live in distinct RabbitMQ instances/servers.
```

Comment:
This was a bit wordy and doesn't actually affect the outcome (still needs to be on its own distinct server), so I've proposed this change:

```suggestion
> Supporting services for an application should not be shared with any other applications at the infrastructure/provisioning level.

For example, a RabbitMQ queue used solely to manage Sidekiq jobs for an application should never be accessed by another app. An exception to the rule would be when services are designed for application communication, such as a RabbitMQ queue that broadcasts publishing events to be consumed by other apps. Both types of queues should live in distinct RabbitMQ instances/servers.
```

I've also made the principle the 'callout', rather than the side note, as it didn't make much sense as a 1-item numbered list.